pkg/host/numa.go

```diff
@@ -0,0 +1,313 @@
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/digitalocean/go-libvirt"
+	"github.com/go-logr/logr"
+	"github.com/ironcore-dev/libvirt-provider/pkg/api"
+	"github.com/ironcore-dev/libvirt-provider/pkg/controllers"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"libvirt.org/go/libvirtxml"
+)
+
+const (
+	KiB  = "KiB"
+	Byte = "Byte"
+)
+
+type NumaTuner struct {
+	Lv           *libvirt.Libvirt
+	HugePageSize uint64
+	CPUTopology  map[int][]int
+}
+
+// CleanBlockedCPUs removes the set of blockedCPUs from the CPUTopology so that Machines are not scheduled on any of the blockedCPUs.
+func CleanBlockedCPUs(topology map[int][]int, blockedCPUs []int) map[int][]int {
+	if len(blockedCPUs) == 0 {
+		return topology
+	}
+
+	blockedSet := sets.New[int]()
+	blockedSet.Insert(blockedCPUs...)
+	for n, cpus := range topology {
+		nodeCPUs := sets.New[int]()
+		nodeCPUs.Insert(cpus...)
+
+		topology[n] = sets.List(nodeCPUs.Difference(blockedSet))
+	}
+
+	return topology
+}
+
+func InitNumaTuneFunc(lv *libvirt.Libvirt, HugePageSize uint64, blockedCPU []int) (controllers.TuneFunc, error) {
+	cpuTopology, err := CpuTopology(lv)
+	if err != nil {
+		return nil, fmt.Errorf("error while fetching the CPU Topology from the host: %w", err)
+	}
+	cpuTopology = CleanBlockedCPUs(cpuTopology, blockedCPU)
+	nt := NumaTuner{
+		Lv:           lv,
+		HugePageSize: HugePageSize,
+		CPUTopology:  cpuTopology,
+	}
+	return nt.Tune, nil
+}
+
+func (n *NumaTuner) Tune(ctx context.Context, domain *libvirtxml.Domain, _ *api.Machine, _ *controllers.MachineReconciler) error {
+	log, err := logr.FromContext(ctx)
+	if err != nil {
+		return fmt.Errorf("can't get logger from context: %w", err)
+	}
+
+	if domain.VCPU == nil || domain.Memory == nil {
+		return nil
+	}
+
+	// TODO: add supporting NUMA-aware pinning without hugepages
+	if domain.MemoryBacking == nil || domain.MemoryBacking.MemoryHugePages == nil {
+		return nil
+	}
+	requiredMemory := uint64(domain.Memory.Value)
+
+	if domain.Memory.Unit != Byte {
+		return fmt.Errorf("unsupported memory unit: %v, please set in bytes", domain.Memory.Unit)
+	}
+
+	pins, err := VMCPUPins(n.Lv)
+	if err != nil {
+		return err
+	}
+
+	log.V(1).Info("Numa aware and hugepage memory tuning", "requiredMemory", requiredMemory, "domain", domain.Name)
+	if ok, err := n.isEnoughFreePagesAllocated(requiredMemory); !ok || err != nil {
+		if !ok {
+			return fmt.Errorf("not enough  memory in  preallocated amount of hugepages ")
+		}
+		return fmt.Errorf("unable to get free pages info: %w", err)
+	}
+
+	allocatableMemoryPerNode, err := n.getFreeNumaNodes(log, requiredMemory)
+	if err != nil {
+		return err
+	}
+
+	nodeCPUs := n.CPUTopology
+	for node, cpus := range nodeCPUs {
+		//	CPU cores with less pinned vm count go first
+		sort.SliceStable(cpus, func(i, j int) bool {
+			return pins[cpus[i]] < pins[cpus[j]]
+		})
+		nodeCPUs[node] = cpus
+	}
+
+	var totalMemory uint64
+	nodeIDs := make([]int, 0, len(allocatableMemoryPerNode))
```

Comment:
optional: you call this len(allocatableMemoryPerNode) 2 times. You can store value into variable and use it on line 127